src/metabase/upload.clj

```diff
@@ -68,8 +67,7 @@
    ::int         ::int-pk})
 
 (def ^:private types
-  (set/union (set (keys type->parent))
-             (set (vals type->parent))))
+  (keys type->parent))
```

Comment:
now you can never have a type without a parent, unless explicitly stated

Reply:
Given that it's only used in a loop a few lines down, does it make sense to have this variable at all? Same question about `pk-base-types`, it would work fine as a map anyway...

Reply:
This is bikeshedding a little but it makes sense to me to keep this variable. For one it allows `type->parent` to be changed to a function without changing `type->ancestors`. 

The second reason is without the `types` variable, the code is objectively harder to understand and more likely to have bugs. If you wanted to understand whether `type->ancestors` contains the key `::text`, you would need to know that `type->parent` also contains the `::text` key. That seems like an unstable assumption (evidence: I just changed it). But with the `types` variable, you just need to understand that `types` contains the full set of types, which includes `::text`. That allows me to freely remove the `::text` variable from `type->parent` again, update the definition of `types`, and not have to worry about anything else breaking.

Reply:
As for `pk-base-types`, the same reason applies: it allows `base-type->pk-type` to be changed to a function without breaking its users.